src/hotspot/cpu/ppc/foreignGlobals_ppc.cpp

```diff
@@ -23,31 +23,236 @@
+      if (to_reg.segment_mask() == REG32_MASK) {
+        __ stfs(as_FloatRegister(from_reg), -8, R1_SP);
+        __ lwa(as_Register(to_reg), -8, R1_SP);
+      } else {
+        __ stfd(as_FloatRegister(from_reg), -8, R1_SP);
+        __ ld(as_Register(to_reg), -8, R1_SP);
+      }
+      break;
+    case StorageType::FLOAT:
+      __ fmr_if_needed(as_FloatRegister(to_reg), as_FloatRegister(from_reg));
+      break;
+    case StorageType::STACK:
+      if (from_reg.segment_mask() == REG32_MASK) {
+        assert(to_reg.stack_size() == 4, "size should match");
+        // TODO: Check if AIX needs 4 Byte offset
+        __ stfs(as_FloatRegister(from_reg), reg2offset(to_reg, out_stk_bias), R1_SP);
+      } else {
+        assert(to_reg.stack_size() == 8, "size should match");
+        __ stfd(as_FloatRegister(from_reg), reg2offset(to_reg, out_stk_bias), R1_SP);
+      }
+      break;
+    default: ShouldNotReachHere();
+  }
+}
+
+static void move_stack(MacroAssembler* masm, Register callerSP, int in_stk_bias, int out_stk_bias,
+                       VMStorage from_reg, VMStorage to_reg) {
+  int out_bias = 0;
+  switch (to_reg.type()) {
+    case StorageType::INTEGER_AND_FLOAT:
+      // "no partial DW rule"
+      __ lfs(F13, reg2offset(from_reg, in_stk_bias), callerSP); // fallthrough
+    case StorageType::INTEGER:
+      switch (from_reg.stack_size()) {
+        case 8: __ ld( as_Register(to_reg), reg2offset(from_reg, in_stk_bias), callerSP); break;
+        case 4: __ lwa(as_Register(to_reg), reg2offset(from_reg, in_stk_bias), callerSP); break;
+        default: ShouldNotReachHere();
+      }
+      break;
+    case StorageType::FLOAT:
+      switch (from_reg.stack_size()) {
+        case 8: __ lfd(as_FloatRegister(to_reg), reg2offset(from_reg, in_stk_bias), callerSP); break;
+        case 4: __ lfs(as_FloatRegister(to_reg), reg2offset(from_reg, in_stk_bias), callerSP); break;
+        default: ShouldNotReachHere();
+      }
+      break;
+    case StorageType::STACK_AND_FLOAT:
+      // "no partial DW rule"
+      __ lfs(F13, reg2offset(from_reg, in_stk_bias), callerSP); // fallthrough
+    case StorageType::STACK:
+      out_bias = out_stk_bias; // fallthrough
+    case StorageType::FRAME_DATA: {
+      switch (from_reg.stack_size()) {
+        case 8: __ ld( R0, reg2offset(from_reg, in_stk_bias), callerSP); break;
+        case 4: __ lwa(R0, reg2offset(from_reg, in_stk_bias), callerSP); break;
+        default: ShouldNotReachHere();
+      }
+      switch (to_reg.stack_size()) {
+        case 8: __ std(R0, reg2offset(to_reg, out_bias), R1_SP); break;
+        case 4: __ stw(R0, reg2offset(to_reg, out_bias), R1_SP); break;
+        default: ShouldNotReachHere();
+      }
+    } break;
+    default: ShouldNotReachHere();
+  }
 }
 
 void ArgumentShuffle::pd_generate(MacroAssembler* masm, VMStorage tmp, int in_stk_bias, int out_stk_bias, const StubLocations& locs) const {
-  Unimplemented();
+  Register callerSP = as_Register(tmp); // preset
```

Comment:
There's no BP register. It could get loaded from the back chain, but would still need a register. R0 is always available as scratch reg, so there's no need to pass it. (Note that I'm not going through stack because of the lack of registers.)
Yeah, maybe passing callerSP separately would make it better readable on PPC64. Not sure if it's worth changing.